inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/service/feature/SemanticSignatureFeatureGenerator.java

```diff
@@ -0,0 +1,233 @@
+        propertyWithLabels = FileUtils.loadPropertyLabels(
+                new File(aRepoProperties.getPath(), "/resources/properties_with_labels.txt"));
+    }
+    
+    @Override
+    public void apply(CandidateEntity aCandidate)
+    {
+        Optional<List<String>> optMentionContext = aCandidate.get(KEY_MENTION_CONTEXT);
+        
+        if (!optMentionContext.isPresent()) {
+            return;
+        }
+        
+        Set<String> mentionContext = new HashSet<>(optMentionContext.get());
+        
+        SemanticSignature sig = getSemanticSignature(aCandidate.getHandle().getKB(),
+                aCandidate.getIRI());
+        Set<String> relatedEntities = sig.getRelatedEntities();
+        Set<String> signatureOverlap = new HashSet<>();
+        for (String entityLabel : relatedEntities) {
+            for (String token : entityLabel.split(" ")) {
+                if (mentionContext.contains(token.toLowerCase(aCandidate.getLocale()))) {
+                    signatureOverlap.add(entityLabel);
+                    break;
+                }
+            }
+        }
+        
+        aCandidate.put(KEY_SIGNATURE_OVERLAP, signatureOverlap);
+        aCandidate.put(KEY_SIGNATURE_OVERLAP_SCORE, signatureOverlap.size());
+        aCandidate.put(KEY_NUM_RELATIONS,
+                (sig.getRelatedRelations() != null) ? sig.getRelatedRelations().size() : 0);
+    }
+    
+    /**
+     * Remove all cache entries of a specific project
+     * @param aEvent
+     *            The event containing the project
+     */
+    @EventListener
+    public void onKnowledgeBaseConfigurationChangedEvent(
+        KnowledgeBaseConfigurationChangedEvent aEvent)
+    {
+        // FIXME instead of maintaining one global cache, we might maintain a cascaded cache
+        // where the top level is the project and then for each project we have sub-caches.
+        // Then we could invalidate only a specific project's cache. However, right now,
+        // we don't have that and there is no way to properly iterate over the caches and
+        // invalidate only entries belonging to a specific project. Thus, we need to
+        // invalidate all.
+        semanticSignatureCache.invalidateAll();
+    }
+    /**
+     * Retrieves the semantic signature of an entity. See documentation of SemanticSignature class.
+     */
+    private SemanticSignature getSemanticSignature(KnowledgeBase aKB, String aIri)
+    {
+        return semanticSignatureCache.get(new SemanticSignatureCacheKey(aKB, aIri));
+    }
+
+    private SemanticSignature loadSemanticSignature(SemanticSignatureCacheKey aKey)
+    {
+        Set<String> relatedRelations = new HashSet<>();
+        Set<String> relatedEntities = new HashSet<>();
+        try (RepositoryConnection conn = kbService.getConnection(aKey.getKnowledgeBase())) {
+            TupleQuery query = SPARQLQueryStore.generateSemanticSignatureQuery(conn,
```

Comment:
This code is really difficult to read, I hope the new sparql builder makes it nicer.

Reply:
Well, it needs to be rewritten and probable the SPARQLQueryBuilder needs to be extended to support the kind of query that is required here.